lean4-infoview/src/infoview/util.ts

```diff
@@ -261,3 +263,61 @@ export function useLogicalDom(ref: React.RefObject<HTMLElement>): [LogicalDomTra
     React.useMemo(() => ({registerDescendant}), [parentCtx])
   ]
 }
+
+type Status = 'pending' | 'fulfilled' | 'rejected'
+
+/** This hook will run the given promise function whenever the deps change.
+ * There will only ever be one in-flight promise at a time. If the deps
+ * change while a promise is still in-flight, then the function will be run
+ * again after the first promise has resolved.
```

Comment:
The purpose of this function is to prevent race conditions where the RPC requests return in a different order than they were sent, right?
1. Request 1 is sent with, say, line=42.
2. Request 2 is sent with line=90.
3. Request 2 returns with diags=[].
4. Request 1 returns with diags=['error'].
And without `useAsync` we would now return the diagnostics for line 42 even though we're at line 90.

If that's the case, then that should be written in the docstring.

Is there a reason for "only ever be one in-flight promise at a time"?

Reply:
The reason for only having one at a time is so that you don't spam server with requests but maybe the server can handle it.

Reply:
Instead of the 'one at a time' model we could instead have the callback of the promise check whether it has been superceded by a later rpc call and discard its result. That would also work.

Reply:
It might be worth it to check if the more eager (multiple requests) approach increases responsiveness or not.

Reply:
I added an unthrottled version, not sure what the best way to test is. They both seem to work.

Reply:
Let's just stick with the unthrottled one for now.